sycl/include/CL/sycl/builtins.hpp

```diff
@@ -27,95 +27,220 @@ namespace __sycl_std = __host_std;
 
 #ifdef __FAST_MATH__
 #define __FAST_MATH_GENFLOAT(T)                                                \
-  (detail::is_genfloatd<T>::value || detail::is_genfloath<T>::value)
+  (detail::is_svgenfloatd<T>::value || detail::is_svgenfloath<T>::value)
 #else
-#define __FAST_MATH_GENFLOAT(T) (detail::is_genfloat<T>::value)
+#define __FAST_MATH_GENFLOAT(T) (detail::is_svgenfloat<T>::value)
 #endif
 
 /* ----------------- 4.13.3 Math functions. ---------------------------------*/
+#define __SYCL_MATH_FUNCTION_OVERLOAD(NAME)                                    \
+  template <typename T, size_t N>                                              \
+  inline __SYCL_ALWAYS_INLINE                                                  \
+      std::enable_if_t<detail::is_sgenfloat<T>::value, sycl::marray<T, N>>     \
+      NAME(sycl::marray<T, N> x) __NOEXC {                                     \
+    sycl::marray<T, N> res;                                                    \
+    auto x_vec2 = reinterpret_cast<sycl::vec<T, 2> const *>(&x);               \
+    auto res_vec2 = reinterpret_cast<sycl::vec<T, 2> *>(&res);                 \
+    for (size_t i = 0; i < N / 2; i++) {                                       \
```

Comment:
This implementation was chosen for simplicity and because it is the most optimal implementation for the CUDA backend for the functions which can use specialized f16x2 instructions that are defined here: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#half-precision-floating-point-instructions. 

It would be possible to choose some different implementations that made use of vectors of size 16, 8, and 4 for example, but, without having a good understanding of how different implementations would affect different backends, I decided that this simple implementation was best, at least as an initial solution. I'm not sure how the logarithmic loop would work in this case, since the largest `sycl::vec` size is 16.

Reply:
I've added a comment to this effect in the code: note that using larger vector sizes could also be better for the CUDA backend: it wouldn't prevent the libclc implementation of the vector cases from using the vector of size two ptx instructions for the operations because this is what happens already for the vector implementations. 
However as I noted here: https://github.com/intel/llvm/pull/6038#discussion_r891673257 the load/store operations are not being vectorized after compilations in the case that we construct the vectors from an marray. I planned to investigate why the compiler is doing this after adding an initial implementation (since I imagined a not perfectly optimized working implementation is better than no implementation), but I can look into why the compiler is not vectorizing the loads/stores now.